src/core/presence.ts

```diff
@@ -350,31 +465,242 @@ export class DefaultPresence
   }
 
   /**
-   * Method to handle and emit presence events
-   * @param member - PresenceMessage ably-js object
-   * @returns void - Emits a transformed event to all subscribers, or upon failure,
-   * the promise will be rejected with an {@link ErrorInfo} object which explains the error.
-   */
-  subscribeToEvents = (member: Ably.PresenceMessage) => {
-    try {
-      // Ably-js never emits the 'absent' event, so we can safely ignore it here.
-      this.emit(member.action as PresenceEvents, {
-        action: member.action as PresenceEvents,
-        clientId: member.clientId,
-        timestamp: member.timestamp,
-        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-        data: member.data?.userCustomData as PresenceData,
+   * Converts a presence message to a presence member, with the action set to 'present'.
+   * @param presenceMessage
+   * @private
+   */
+  private _presenceMessageToPresenceMember(presenceMessage: Ably.PresenceMessage): InternalPresenceMember {
+    this._logger.trace('Presence._presenceMessageToPresenceMember();', { presenceMessage });
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const { clientId, timestamp, extras, data, connectionId } = presenceMessage;
+    const chatPresenceData = data as ChatPresenceData;
+    return {
+      connectionId,
+      clientId,
+      action: 'present' as PresenceEvents,
+      data: chatPresenceData.userCustomData,
+      updatedAt: timestamp,
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      extras: extras,
+      _nonce: chatPresenceData.presence?.nonce,
+    };
+  }
+
+  /**
+   * Sets the presence members from the provided presence messages.
+   * @param messages
+   * @private
+   */
+  private _setPresenceMembersFromMessages(messages: PresenceMessage[]): void {
+    this._logger.trace('Presence._setPresenceMembersFromMessages();', { messages });
+    // Clear the current presence members
+    this._presenceMembers.clear();
+    for (const message of messages) {
+      this._setPresenceMember(message);
+    }
+    this._logger.trace('Presence._setPresenceMembersFromMessages(); New presence members set', {
+      members: [...this._presenceMembers.values()],
```

Comment:
Do we need to log all of this? Perhaps just the length? Quite expensive to log on every change